pkg/e2e/verify/nodepool.go

```diff
@@ -0,0 +1,474 @@
+		fmt.Sprintf("Node %s missing AWS provider ID", node.Name))
+
+	hasInternalIP := false
+	for _, addr := range node.Status.Addresses {
+		if addr.Type == corev1.NodeInternalIP && addr.Address != "" {
+			hasInternalIP = true
+			break
+		}
+	}
+	Expect(hasInternalIP).To(BeTrue(), fmt.Sprintf("Node %s missing InternalIP", node.Name))
+
+	Expect(node.Labels["node.kubernetes.io/instance-type"]).ToNot(BeEmpty(),
+		fmt.Sprintf("Node %s missing instance type", node.Name))
+}
+
+func getClusterNamespace(h *helper.H) (string, error) {
+	if name := viper.GetString(config.Cluster.Name); name != "" {
+		return fmt.Sprintf("clusters-%s", name), nil
+	}
+
+	current := h.CurrentProject()
+	if strings.HasPrefix(current, "clusters-") {
+		return current, nil
+	}
+
+	return "", fmt.Errorf("could not determine cluster namespace")
+}
+
+func canAccessNodePoolAPI(h *helper.H, namespace string) bool {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	_, err := h.Dynamic().Resource(nodePoolGVR).Namespace(namespace).List(ctx, metav1.ListOptions{Limit: 1})
+	return err == nil
+}
+
+func buildNodePoolSpec(name, namespace string) map[string]interface{} {
+	clusterName := strings.TrimPrefix(namespace, "clusters-")
+
+	return map[string]interface{}{
+		"apiVersion": "hypershift.openshift.io/v1beta1",
+		"kind":       "NodePool",
+		"metadata": map[string]interface{}{
+			"name":      name,
+			"namespace": namespace,
+		},
+		"spec": map[string]interface{}{
+			"clusterName": clusterName,
+			"replicas":    1,
+			"management": map[string]interface{}{
+				"autoRepair":  true,
+				"upgradeType": "Replace",
+			},
+			"platform": map[string]interface{}{
+				awsProvider: map[string]interface{}{
+					"instanceType": "m5.large",
+				},
+			},
+		},
+	}
+}
+
+func waitForNodePoolReady(ctx context.Context, h *helper.H, namespace, nodePoolName string, timeout time.Duration) error {
+	return wait.For(func(ctx context.Context) (bool, error) {
+		np, err := h.Dynamic().Resource(nodePoolGVR).Namespace(namespace).
+			Get(ctx, nodePoolName, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+
+		ready, found, err := unstructured.NestedInt64(np.Object, "status", "readyReplicas")
+		if err != nil || !found {
+			return false, nil
+		}
+
+		return ready >= 1, nil
+	}, wait.WithTimeout(timeout), wait.WithInterval(30*time.Second))
+}
+
+func failIfSTSError(err error, operation, permissions string) {
+	if err == nil {
+		return
+	}
+	msg := strings.ToLower(err.Error())
+	patterns := []string{
+		"accessdenied", "unauthorizedoperation", "forbidden",
+		"invalid iam role", "sts permissions", "not authorized",
+		"assumerolewithwebidentity", "invalididentitytoken",
+		"authfailure", "signaturedoesnotmatch",
+	}
+	for _, pattern := range patterns {
+		if strings.Contains(msg, pattern) {
+			ginkgo.Fail(fmt.Sprintf(
+				"STS_PERMISSION_ERROR: %s failed due to missing AWS permissions (%s): %v. "+
```

Comment:
This should be cleaned up.  There's no need to concat strings within a `fmt.Sprintf` method call.